runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactStagingService.java

```diff
@@ -0,0 +1,292 @@
+        proxyManifestBuilder.addLocation(Location.newBuilder()
+            .setName(artifactMetadata.getName())
+            .setUri(artifactDirResourceId
+                .resolve(encodedFileName(artifactMetadata), StandardResolveOptions.RESOLVE_FILE)
+                .toString()).build());
+      }
+      try (WritableByteChannel manifestWritableByteChannel = FileSystems
+          .create(manifestResourceId, MimeTypes.TEXT)) {
+        manifestWritableByteChannel
+            .write(CHARSET.encode(JsonFormat.printer().print(proxyManifestBuilder.build())));
+      }
+      // TODO: Validate integrity of staged files.
+      responseObserver.onNext(CommitManifestResponse.newBuilder()
+          .setStagingToken(manifestResourceId.toString())
+          .build());
+      responseObserver.onCompleted();
+    } catch (IOException e) {
+      // TODO: Cleanup all the artifacts.
+      LOG.error("Unable to commit manifest.", e);
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Nothing to close here.
+  }
+
+  /**
+   * Generate a stagingSessionToken compatible with {@link BeamFileSystemArtifactStagingService}.
+   *
+   * @param sessionId Unique sessionId for artifact staging.
+   * @param basePath Base path to upload artifacts.
+   * @return Encoded stagingSessionToken.
+   */
+  public static String generateStagingSessionToken(String sessionId, String basePath)
+      throws Exception {
+    StagingSessionToken stagingSessionToken = new StagingSessionToken();
+    stagingSessionToken.setSessionId(sessionId);
+    stagingSessionToken.setBasePath(basePath);
+    return encodeStagingSessionToken(stagingSessionToken);
+  }
+
+  private String encodedFileName(ArtifactMetadata artifactMetadata) {
+    return "artifact_" + Hashing.sha256().hashString(artifactMetadata.getName(), CHARSET)
+        .toString();
+  }
+
+  private StagingSessionToken decodeStagingSessionToken(String stagingSessionToken)
+      throws IOException {
+    try {
+      return MAPPER.readValue(stagingSessionToken, StagingSessionToken.class);
+    } catch (IOException e) {
+      try {
+        LOG.error(
+            "Unable to deserialize staging token {}. Expected format {}. Error {}",
+            stagingSessionToken, MAPPER.writeValueAsString(new StagingSessionToken()),
+            e.getMessage());
+      } catch (JsonProcessingException e1) {
+        LOG.error("Error {} occurred while serializing {}.", e.getMessage(),
+            StagingSessionToken.class);
+      }
+      throw e;
+    }
+  }
+
+  private static String encodeStagingSessionToken(StagingSessionToken stagingSessionToken)
+      throws Exception {
+    try {
+      return MAPPER.writeValueAsString(stagingSessionToken);
+    } catch (JsonProcessingException e) {
+      LOG.error("Error {} occurred while serializing {}.", e.getMessage(),
+          StagingSessionToken.class);
```

Comment:
This error seems very unlikely to happen, but it would be helpful to include the actual value of the token (.toString(), and add a reasonable implementation of toString() to it)

Reply:
Make sense.